File: src-tauri/src/handlers/diagnosis.rs
use crate::datastore::diagnosis;
use crate::types::{Diagnoses, Diagnosis};

#[tauri::command]
pub async fn add_diagnoses(
    patient_id: String,
    diagnoses: Vec<Diagnosis>,
    date: String,
    window: tauri::Window,
) -> Result<String, String> {
    diagnosis::add_diagnoses_db(patient_id, diagnoses, date, &window)
}

#[tauri::command]
pub async fn get_diagnoses_by_id(
    diagnoses_id: String,
    window: tauri::Window,
) -> Result<Diagnoses, String> {
    diagnosis::get_diagnoses_by_id_db(diagnoses_id, &window)
}

#[tauri::command]
pub async fn update_diagnoses(diagnoses: Diagnoses, window: tauri::Window) -> Result<(), String> {
    diagnosis::update_diagnoses_db(diagnoses, &window)
}

#[tauri::command]
pub async fn get_all_diagnoses_by_patient_id(
    patient_id: String,
    window: tauri::Window,
) -> Result<Vec<Diagnoses>, String> {
    diagnosis::get_all_diagnoses_by_patient_id_db(patient_id, &window)
}

File: src/api/diagnosis.ts
import { invoke } from "@tauri-apps/api/core";
import { TDiagnosis, TDiagnoses } from "../types";

export const getDiagnosesByIdApi = async (diagnosesId: string) => {
  try {
    const res = await invoke<TDiagnoses>("get_diagnoses_by_id", {
      diagnosesId,
    });
    return res;
  } catch (e) {
    console.error("Error getting diagnoses:", e);
  }
};

export const addDiagnosesApi = async (
  diagnosesArray: TDiagnosis[],
  patientId: string,
) => {
  try {
    const res = await invoke<string>("add_diagnoses", {
      patientId,
      diagnoses: diagnosesArray,
      date: new Date().toISOString(),
    });
    return res;
  } catch (e) {
    console.error("Error adding diagnoses:", e);
  }
};

export const getAllDiagnosesByPatientIdApi = async (patientId: string) => {
  try {
    const res = await invoke<TDiagnoses[]>("get_all_diagnoses_by_patient_id", {
      patientId,
    });
    return res;
  } catch (e) {
    console.error("Error getting patient diagnoses:", e);
    return [];
  }
};

export const updateDiagnosesApi = async (diagnoses: TDiagnoses) => {
  try {
    await invoke<void>("update_diagnoses", {
      diagnoses,
    });
    return true;
  } catch (e) {
    console.error("Error updating diagnoses:", e);
    return false;
  }
};
export const updateDiagnosisApi = async (
  diagnosesId: string,
  diagnosisId: string,
  diagnosisData: {
    diagnosis_title: string;
    diagnosis_type: string;
    start: string;
    end: string;
    comment: string;
  },
) => {
  try {
    // Step 1: Get the complete current diagnoses object
    const diagnoses = await getDiagnosesByIdApi(diagnosesId);

    if (!diagnoses) {
      throw new Error("Diagnoses not found");
    }

    // Step 2: Update the specific diagnosis in the array
    if (diagnoses.diagnoses) {
      diagnoses.diagnoses = diagnoses.diagnoses.map((diagnosis) => {
        if (diagnosis.id === diagnosisId) {
          return {
            ...diagnosis,
            diagnosis_title: diagnosisData.diagnosis_title,
            diagnosis_type: diagnosisData.diagnosis_type,
            start: diagnosisData.start,
            end: diagnosisData.end,
            comment: diagnosisData.comment,
          };
        }
        return diagnosis;
      });
    }

    // Step 3: Update the timestamp
    diagnoses.date = new Date().toISOString();

    // Step 4: Call the existing updateDiagnosesApi to save to backend
    return await updateDiagnosesApi(diagnoses);
  } catch (e) {
    console.error("Error updating diagnosis:", e);
    return false;
  }
};
